### ISSUE TITLE
Interactive Reconfigure PR

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#547]
Consider the case when someone has already activated Renovate but wants to make a configuration change. My idea:

They create an issue "Reconfigure Renovate".
Renovate bot looks for this each run
If found, it opens a "Reconfigure Renovate" PR to close that issue. 
PR is initially "empty" but user can edit renovate.json in branch and see results just like the onboarding PR. 

Essentially, it's an interactive way to update renovate configuration and helps them validate the config.

**Comment by rarkins** [comment]
Problem: only administrators should be able to do this

**Comment by ikatyang** [comment]
How about using a label like `renovate-reconfigure`?

**Comment by rarkins** [comment]
> How about using a label like renovate-reconfigure?

@ikatyang that might be a good solution! So then the issue could be named anything you like (e.g. "aws-sdk updates are too frequent") and then once they add a `renovate-configure` label then that is the trigger to create a matching PR?

**Comment by charlike** [comment]
I believe it is not needed, since you can just edit the `renovate` config (the file or from package.json) from github UI. Renovate already works that way - i tried it. Don't see value to open an issue.

**Comment by rarkins** [comment]
Yes it that means you need to "try and fail" on your `master` branch. It just take someone 2-3 tries to get it right and would be much nicer to have it interactive and can improve as we improve onboarding too.

**Comment by charlike** [comment]
mmm probably yea, it takes hours. it only will make sense if opening an issue makes that process faster, which means that renovate should install some listener for such named issues

**Comment by rarkins** [comment]
It will not make it faster - yet - but it will make it easier and less risky immediately.

**Comment by charlike** [comment]
I don't see how it may be useful if it takes same hours (actually, it is not hours - that depens on schedule config), but anyway.

In anyway, maybe it's good idea to think about setting some listener for faster triggering of the runs. I was thinking about that exactly because testing options and behaviors. Is it possible to set `schedule` option to something like `every 30 minutes in weekday` or even 10, haha? I've seen that this is based on `later` lib, but didn't dive deeply there.

Another reason because setting a listener (probably webhook?) is good, is that because when i ran `renovate` cli, it just created branches (after i accepted the Configure PR), but didn't created the PRs for these branches - even after hours, don't know probably a bug?

**Comment by rarkins** [comment]
Speed is not the only form of usefulness that exists. I don't agree with your conclusion of "unless this speeds things up then I don't see the benefit". There would be many who are unsure if their configuration changes are correct or not and prefer to test in a branch rather than hacking away on `master` until it's right.

Importantly this doesn't force you to do anything differently so if you still don't see the benefits this would provide to a non "power user" then it's ok because you can keep committing your `renovate.json` directly as you were before. 

Finally, let's not discuss other topics (scheduling) in the wrong issue. Please raise a new issue and we can discuss there.

**Comment by rarkins** [comment]
Waiting on #1399

**Comment by zephraph** [comment]
I'd mentioned an idea in #2879 along these same lines. It'd be nice to not have to have the over head of a special label or issue. If it was possible for renovate to listen for changes to the relevant configuration files on each PR, it could update it based on that info. 

Granted, unless renovate is already listening to changes on every PR I know that's a potentially a lot more computational overhead.  Thoughts?

**Comment by rarkins** [comment]
Renovate does already look for PRs that change a renovate configuration file, so this is not out of the question. To reduce the burden, we only look at ones that have "renovate" in the PR title or branch name (case insensitive check). The bigger challenge was how to run a "mock run" with the branch's config and display it - it would be like Renovate running inside Renovate, and perhaps even more than one time if there was more than one PR altering renovate config.

**Comment by zephraph** [comment]
Yeah, that definitely sounds challenging. I was taking a look at dry run as a valid alternative in the short term. It's unclear to me if you can run the renovate cli on a particular branch in a repo or if it only ever runs on master.

**Comment by rarkins** [comment]
I've never tried it but you might be able to do a dry-run CLI that way by specifying your reconfigure branch as `baseBranches`

**Comment by klieber** [comment]
I was about to create a feature request for something very similar to what @zephraph request in #2879.  I simply just want to extend the process where renovate validates any PR that contains changes to renovate configuration so that it also updates the PR with what new pull requests it would add if I merged my configuration changes.  

My use case is that we just started using renovate so we had a lot of dependencies that needed upgraded and it was going to create around 75 pull requests so we configured renovate to disable a lot of the dependencies that we considered risky upgrades.  That left us with around 40 pull requests when we finished onboarding.  Now those 40 have all been merged and I want to start enabling those dependencies we originally disabled but I want to review them all first before having renovate create the PRs.  

I know I can also rename the onboarding pull request to force onboarding again but that will also stop renovate from keeping other dependencies updated.  I might not be ready to merge in the configuration changes for another 2 or 3 weeks so I don't want to completely disable renovate during that time.

Also, on the idea of having an issue or a label trigger renovate.  Remember that not all platforms will support that.  For instance, we are using Bitbucket Server and we have no way to create "Issues" within Bitbucket Server. Our issue management system is JIRA.

**Comment by jamesdh** [comment]
The [docs](https://docs.renovatebot.com/reconfigure-renovate/) indicate this feature should exist but in my testing it's not working as described and this PR makes me wonder if it's actually supposed to be working. Help?

**Comment by zephraph** [comment]
@klieber one option is that you could enable those packages but have them behind the [`masterIssueApproval`](https://docs.renovatebot.com/configuration-options/#masterissueapproval) flag. 

Here's a snippet from our [config](https://github.com/artsy/renovate-config/blob/master/lib/config.js#L21-L27):
  
```js
{
  managers: ["npm"],
  depTypeList: ["dependencies", "devDependencies", "peerDependencies"],
  packagePatterns: ["*"],
  excludePackagePatterns: ["^@artsy", "^@omakase"],
  masterIssueApproval: true
}
```

This is a [package rule](https://docs.renovatebot.com/configuration-options/#packagerules) that enables `massterIssueApproval` for all dependencies which aren't in the scope `@artsy` or `@omakase`. It'll create an [issue](https://github.com/artsy/reaction/issues/2400) where all your dependency updates would be stored...

That said, it's really a work around. If you messed up, you could still get spammed with a bunch of PRs which you didn't intend. @rarkins, a more official solution would still be nice.

**Comment by rarkins** [comment]
I was going to recommend `masterIssueApproval` too actually. If you set it at the top level (i.e. not within a package rule) then there should be no possibility of getting "spammed with a bunch of PRs".

Also remember that Renovate's default setting within `config:base` is maximum 2 PRs per hour so even with a mistake you should notice you are getting PRs you didn't intend to.

Considering that the master issue gives a good preview of PRs, it's unlikely that this issue (Interactive Reconfigure PR) will receive any priority - it would be a lot of work to implement.

**Comment by ben-foxmoore** [comment]
> Renovate does already look for PRs that change a renovate configuration file, so this is not out of the question. Currently it tries to verify the config is valid so you don't merge invalid config to master.

@rarkins can you expand on this? I'm struggling to find anything about it in the documentation. Is it still there?

**Comment by rarkins** [comment]
Renovate no longer does that. U need to run the renovate-config-validator yourself, which is bundled with renovate

**Comment by ben-foxmoore** [comment]
Got it - thanks!

**Comment by karfau** [comment]
I would love to understand why those checks were disabled. I think this would actually prevent the issue of what I described in # all together.
Maybe it could be put behind some opt in configuration that reduces the related work?

Here is what I was thinking about over there:
The idea of opening an issue in the repo where change is needed might be nice, but maybe there could also be other "entry points" that are closer to the action somebody would naturally do: like creating a branch (starting with the configured branch prefix?) that makes a change to renovate.json, or creating a PR mentioning the renovatebot? Or even adding a commit to any branch but the default one that mentions the bot?

Or, following the idea of the "initially empty PR": having a checkbox in the dependency dashboard, the original onboarding PR or any of the created PRs that can be checked to create such a branch/PR.

My comment might be a bit off topic, but still wanted to add it here, since it's very related.

**Comment by rarkins** [comment]
Needing to fetch every open PR from the API just in case one of them changed Renovate's config was non-performant. It was easier for us to just fetch every open and closed PR created by our account as a single query. However if the topic can be revisited and an efficient approach found then it could be added back.

**Comment by viceice** [comment]
I think a better approach would be (at least for github actions) to use our github action to validate with a worflow, this needs renovatebot/github-action#548

I've a github pipeline where i use npx to validate 8 configs in ~1 minute (separate jobs)

**Comment by reitzig** [comment]
FWIW, I for one would be totally fine requesting a dry-run via (web-)hook from either the Git server (in my case: Bitbucket) or the CI pipeline (in my case: Jenkins).

What I totally see happening is having to configure a slew of regex matchers -- which will never be correct on the first or twentieth attempt. I don't want to do that kind of debugging on the main branch, and I don't want to wait for the hourly re-run of Renovate.

Currently, the only workaround I see is to start a dry-run (for only the current repo and branch) myself as part of the pipeline; any issues would fail the build (fine) and devs would have to read the raw log (meh, but okay).

Ideally, Renovate would be able to create a comment similar to the onboarding one on my PR (of a branch that changes the config file), and keep it current. Knowing repository and branch (see above), at least the problem of _finding_ the PR to work with should be feasible.

**Comment by rarkins** [comment]
Related/blocked by: #13948

**Comment by MPV** [comment]
Since this has been implemented now:
- https://github.com/renovatebot/renovate/pull/24699

...could we review/reconsider this again now? ❤️

**Comment by rarkins** [comment]
PRs are welcome, this feature is no longer blocked

**Comment by RahulGautamSingh** [comment]
There will need to be a slight shift in behavior of the reconfigure pr.

We should ask users to only create the branch and not the PR from now on, as we cannot change the pr body of a PR from another user, as the bot token won't have enough permissions for that.

So, users should create a branch and wait. If validation passes, a PR will be created automatically.

**Comment by rarkins** [comment]
Can the bot add comments?

**Comment by RahulGautamSingh** [comment]
Yes

**Comment by rarkins** [comment]
So we can do the feedback part using ensureComment?

**Comment by RahulGautamSingh** [comment]
Yes, working on that.

**Comment by RahulGautamSingh** [comment]
I was thinking that we should let the user create the PR and not create it ourselves.

Because, sometimes the user might not want a pr explaining the result of the config-changes and only wabnt to validate the config, which can be achieved by just creating the reconfigure branch.

If they want detailed info about the changes new config will bring then they can open a PR and the bot will add a comment on it.

So the flow will be like:
1. Want to validate reconfigure config? Create a reconfigure branch
2. Want to nkow the result of the config cheanges? Create a PR with sourceBranch = reconfigure branch

This will keep the logic and implementation simpler as we won't need to handle switch between updating pr body and pr comment depending on whether the pr was created by the bot or the user.

**Comment by viceice** [comment]
this is a problem because searching that PR is expensive. by default we've only our own PRs cached locally.

**Comment by rarkins** [comment]
We can know that there's a branch with the specific name first, and if so then have a new query to get just the corresponding PR? Some SCMs probably make it possible in 1 query and then we can also cache that mapping between runs

**Comment by RahulGautamSingh** [comment]
> We can know that there's a branch with the specific name first, and if so then have a new query to get just the corresponding PR?

Current flow:
Firs tcheck the branch exists,

**Comment by RahulGautamSingh** [comment]
Yes, we confirm that branch exists or not and only then look for the PR corresponding to that branch. For github we cache PRs created by other authors as well, in the repo cache. But, this is not implemented for all platforms.

**Comment by RahulGautamSingh** [comment]
Demo Run Result: https://github.com/Rahul-renovate-testing/recon-pr/pull/1 🏍️

**Comment by RahulGautamSingh** [comment]
We use findPr with `includeOtherAuthors=true` for fetching the reconfigure pr

**Comment by MPV** [comment]
A parallel idea (would affect the above):
- https://github.com/renovatebot/renovate/discussions/34651

**Comment by chrko** [comment]
Just another thing for this feature: (Unfortunately) we have quite some restrictions regarding branch names, especially their lengths. We already using therefore the `hashedBranchLength` option. The current branch name for the reconfigure branch is hard-coded. I would like to this also be exposed as an config option, likely only for self-hosted instances.

**Comment by rarkins** [comment]
We will not overcomplicate this already complicated feature initially. @chrko please submit a new discussion idea. You'll probably need to create the PR yourself or wait for someone else with the same requirement
